pkgs/build-support/rust/rust-utils.nix

```diff
@@ -0,0 +1,332 @@
+             echo "-L $L" >> target/link.final
+           done
+         fi
+      done
+      echo "$EXTRA_LINK" | while read i; do
+         if [ ! -z "$i" ]; then
+           for lib in $i; do
+             echo "-l $lib" >> target/link
+             echo "-l $lib" >> target/link.final
+           done
+         fi
+      done
+
+      if [ -e target/link ]; then
+         sort -u target/link.final > target/link.final.sorted
+         mv target/link.final.sorted target/link.final
+         sort -u target/link > target/link.sorted
+         mv target/link.sorted target/link
+
+         tr '\n' ' ' < target/link > target/link_
+         LINK=$(cat target/link_)
+       fi
+
+      mkdir -p target/bin
+      echo "${crateBin}" | sed -n 1'p' | tr ',' '\n' | while read BIN; do
+         if [ ! -z "$BIN" ]; then
+           echo "$boldgreen""Building $BIN$norm"
+           if ${verboseBuild}; then
+             echo "$boldgreen""Running$norm rustc --crate-name $BIN --crate-type bin ${rustcOpts} ${crateFeatures} --out-dir target/bin --emit=dep-info,link -L dependency=target/deps $LINK ${deps}$EXTRA_LIB --cap-lints allow $BUILD_OUT_DIR $EXTRA_BUILD $EXTRA_FEATURES"
+           fi
+           rustc --crate-name $BIN --crate-type bin ${rustcOpts} ${crateFeatures} \
+             --out-dir target/bin --emit=dep-info,link -L dependency=target/deps \
+             $LINK ${deps}$EXTRA_LIB --cap-lints allow \
+             $BUILD_OUT_DIR $EXTRA_BUILD $EXTRA_FEATURES --color ${colors}
+         fi
+      done
+      if [[ (-z "${crateBin}") && (-e src/main.rs) ]]; then
+         echo "$boldgreen""Building src/main.rs (${crateName})$norm"
+         if ${verboseBuild}; then
+           echo "$boldgreen""Running$norm rustc --crate-name $CRATE_NAME src/main.rs --crate-type bin ${rustcOpts} ${crateFeatures} --out-dir target/bin --emit=dep-info,link -L dependency=target/deps $LINK ${deps}$EXTRA_LIB --cap-lints allow $BUILD_OUT_DIR $EXTRA_BUILD $EXTRA_FEATURES"
+         fi
+         rustc --crate-name $CRATE_NAME src/main.rs --crate-type bin ${rustcOpts} \
+           ${crateFeatures} --out-dir target/bin --emit=dep-info,link \
+           -L dependency=target/deps $LINK ${deps}$EXTRA_LIB --cap-lints allow \
+           $BUILD_OUT_DIR $EXTRA_BUILD $EXTRA_FEATURES --color ${colors}
+      fi
+      # Remove object files to avoid "wrong ELF type"
+      rm -f $(find target -name "*.o")
+    '' + finalBins;
+
+    installCrate = crateName: ''
+      mkdir -p $out
+      if [ -s target/env ]; then
+        cp target/env $out/env
+      fi
+      if [ -s target/link.final ]; then
+        cp target/link.final $out/link
+      fi
+      cp target/deps/* $out # */
+      if [ "$(ls -A target/build)" ]; then # */
+        cp -r target/build/* $out # */
+      fi
+      if [ "$(ls -A target/bin)" ]; then # */
+        mkdir -p $out/bin
+        cp -P target/bin/* $out/bin # */
+      fi
+    '';
+in
+
+crate: rust: stdenv.mkDerivation rec {
```

Comment:
I see why you have added these parameter (to use different rust versions). But should this not be also possible
by using `mkRustCrate.override ({rust = rust_foo;})?

Reply:
I don't believe many Rust users use the default Rust from nixpkgs (of course, I might be wrong), since the overlay is much more convenient. What are the pros and cons of doing this?

Reply:
It is the documented way to override dependencies in derivations. If you want your own version of this function, you would do:

```nix
my-rust-crates = callPackage ../my-rust-crates {
  mkRustCrate = mkRustCrate.override ({rust = rustNightly;})
};
```

Reply:
This would be useful as there are some features only available on `rustNightly`.

Reply:
Done! It is actually more elegant (even though I had to modify all my projects' default.nix).